### PR TITLE
[Feat](nereids) add max/min filter push down rewrite rule

### PR DESCRIPTION
**Pull request by feiniaofeiafei**
add rewrite rule:
select id, max(a) from t group by id having max(a)>10;
 ->
 select id, max(a) from t where a>10 group by id;
select id, min(a) from t group by id having min(a)<10;
->
 select id, min(a) from t where a<10 group by id;

### CODE REVIEW

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run cloud_p1

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDownTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Test
+    public void testMinRewrite() {
+        String sql = "select id, min(score) from max_t group by id having min(score)<10";
+        PlanChecker.from(connectContext).analyze(sql).rewrite()
+                .matches(logicalFilter(logicalOlapScan()).when(filter -> filter.getConjuncts().size() == 1));
+    }
+
+    @Test
+    public void testMaxNotRewrite0() {
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDownTest.java:54]: could u give it a more meaningful name?

File: fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDownTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    @Test
+    public void testMinNotRewrite1() {
+        String sql = "select id, min(score) from max_t group by id having min(score)>10";
+        PlanChecker.from(connectContext).analyze(sql).rewrite()
+                .nonMatch(logicalFilter(logicalOlapScan()));
+    }
+
+    @Test
+    public void testMinNotRewrite2() {
+        String sql = "select id, min(score), max(score) from max_t group by id having min(score)>10";
```
Comment (morrySnow) [inline fe/fe-core/src/test/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDownTest.java:69]: test min(a), max(b) and min(a), min(b)?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDown.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+    private Plan pushDownMaxMinFilter(LogicalFilter<LogicalAggregate<Plan>> filter) {
+        Set<Expression> conjuncts = filter.getConjuncts();
+        LogicalAggregate<Plan> agg = filter.child();
+        Plan aggChild = agg.child();
+        List<NamedExpression> aggOutputExpressions = agg.getOutputExpressions();
+        Set<Expression> aggFuncs = ExpressionUtils.collect(aggOutputExpressions,
+                expr -> expr instanceof AggregateFunction);
+        Set<Expression> maxMinFunc = ExpressionUtils.collect(aggFuncs,
+                expr -> expr instanceof Max || expr instanceof Min);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDown.java:79]: if group by key is empty, we should keep the filter upper aggregate

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDown.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                newPredicate = new GreaterThanEqual(func.child(0), originConjunct.child(1));
+            }
+        } else {
+            if (originConjunct instanceof LessThan) {
+                newPredicate = new LessThan(func.child(0), originConjunct.child(1));
+            } else if (originConjunct instanceof LessThanEqual) {
+                newPredicate = new LessThanEqual(func.child(0), originConjunct.child(1));
+            }
+        }
+        Preconditions.checkState(newPredicate != null);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/MaxMinFilterPushDown.java:113]: add error msg

**Comment by feiniaofeiafei** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

File: regression-test/suites/nereids_rules_p0/max_min_filter_push_down/max_min_filter_push_down.groovy [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+// specific language governing permissions and limitations
+// under the License.
+suite("max_min_filter_push_down") {
+    sql "SET enable_nereids_planner=true"
+    sql "SET enable_fallback_to_original_planner=false"
+    sql "SET ignore_shape_nodes='PhysicalDistribute,PhysicalProject'"
+
+    sql "drop table if exists max_min_filter_push_down1"
+    sql"""
+    CREATE TABLE max_min_filter_push_down1 (
```
Comment (xzj7019) [inline regression-test/suites/nereids_rules_p0/max_min_filter_push_down/max_min_filter_push_down.groovy:24]: seems no negative cases, such as min > or max <, or cases with join, original filter, etc.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java [hunk 6]
```diff
@@ -180,7 +181,8 @@ public class Rewriter extends AbstractBatchJobExecutor {
                         topDown(
                                 // ExtractSingleTableExpressionFromDisjunction conflict to InPredicateToEqualToRule
                                 // in the ExpressionNormalization, so must invoke in another job, otherwise dead loop.
-                                new ExtractSingleTableExpressionFromDisjunction()
+                                new ExtractSingleTableExpressionFromDisjunction(),
+                                new MaxMinFilterPushDown()
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/executor/Rewriter.java:185]: Is this the best place?

File: regression-test/suites/nereids_rules_p0/max_min_filter_push_down/max_min_filter_push_down.groovy [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    select max(value1) from max_min_filter_push_down1 having 40<max(value1); 
+    """
+
+    qt_min_equal_scalar_agg_res """
+    select min(value1) from max_min_filter_push_down1 having min(value1) <=40 and min(value1) <=20;
+    """
+    qt_max_equal_scalar_agg_res """
+    select max(value1) from max_min_filter_push_down1 having max(value1) >=40;
+    """
+
```
Comment (xzj7019) [inline regression-test/suites/nereids_rules_p0/max_min_filter_push_down/max_min_filter_push_down.groovy:170]: cases whose literal is a composed expression which can be folded?

**Review by xzj7019 — APPROVED** [review]

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run external

**Review by englefly — APPROVED** [review]

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by feiniaofeiafei** [comment]
run p0

**Comment by feiniaofeiafei** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**